dev-tools/xtask/src/a4x2_package.rs

```diff
@@ -0,0 +1,382 @@
+    cmd!(sh, "banner testbed").run()?;
+    let testbed_dir = env.work_dir.join("testbed");
+    let a4x2_dir = testbed_dir.join("a4x2");
+
+    // TODO move this clone & build to prepare, so that if the omicron build
+    // takes awhile we wont have lost our git token when we get here. Relevant
+    // for CI but not for running locally.
+    cmd!(
+        sh,
+        "git clone https://github.com/oxidecomputer/testbed {testbed_dir}"
+    )
+    .run()?;
+
+    // XXX use main branch
+    // cmd!(sh, "git -C {testbed_dir} checkout artemis/init-in-smf").run()?;
+
+    let _popdir = sh.push_dir(&a4x2_dir);
+
+    // build a4x2
+    let cargo = &env.cargo;
+    cmd!(sh, "{cargo} build --release").run()?;
+
+    sh.copy_file("../target/release/a4x2", &env.out_dir)?;
+
+    // This will modify some omicron configs, build omicron, and package things
+    // up for the a4x2 cargo-bay
+    cmd!(sh, "./config/build-packages.sh")
+        .env("OMICRON", &env.omicron_dir)
+        .run()?;
+
+    // debatable whether this should happen in package or deploy
+    cmd!(sh, "./config/fetch-softnpu-artifacts.sh").run()?;
+
+    // Deduplicate cargo-bay output amongst g0-g3. They differ on a few
+    // files; deduping the rest is critical to keeping the final archive
+    // size manageable for CI. To do this we will
+    // - iterate files in g0
+    // - check equivalent paths for g1/g3
+    // - if hashes match, extract out to `sled-common/`
+    let cargo_bay_dir = a4x2_dir.join("cargo-bay");
+    let common_dir = cargo_bay_dir.join("sled-common");
+    let g0_dir = cargo_bay_dir.join("g0");
+    let prefixes = [
+        cargo_bay_dir.join("g1"),
+        cargo_bay_dir.join("g2"),
+        cargo_bay_dir.join("g3"),
+    ];
+    for ent in WalkDir::new(&g0_dir) {
+        let ent = ent?;
+        if !ent.file_type().is_file() {
+            continue;
+        }
+
+        let g0_path =
+            Utf8PathBuf::from_path_buf(ent.path().to_path_buf()).unwrap();
+        let g0_hash = sha256(&g0_path)?;
+
+        let base_path = g0_path.strip_prefix(&g0_dir)?;
+        let mut can_dedupe = true;
+        for prefix in &prefixes {
+            let path = prefix.join(base_path);
+            // if it doesn't exist in all bays, we can't dedupe
+            if !path.try_exists()? {
+                can_dedupe = false;
+                break;
+            }
+
+            let hash = sha256(&path)?;
+            if hash != g0_hash {
+                can_dedupe = false;
+                break;
+            }
+        }
+
+        if can_dedupe {
+            // yay, extract out to the common dir.
+            let dest = common_dir.join(base_path);
+            sh.create_dir(&dest.parent().unwrap())?;
+            sh.copy_file(&g0_path, &dest)?;
+            sh.remove_path(&g0_path)?;
+            for prefix in &prefixes {
+                sh.remove_path(&prefix.join(base_path))?;
+            }
+        }
```

Comment:
Hmm interesting -- what code ensures that sled-common is checked?

Reply:
In a4x2-deploy, I've got this block of commands

```
    cmd!(sh, "tar -cf sled-common.tar -C {a4x2_dir}/cargo-bay/sled-common ./")
        .run()?;
    cmd!(sh, "tar -xf sled-common.tar -C {a4x2_dir}/cargo-bay/g0").run()?;
    cmd!(sh, "tar -xf sled-common.tar -C {a4x2_dir}/cargo-bay/g1").run()?;
    cmd!(sh, "tar -xf sled-common.tar -C {a4x2_dir}/cargo-bay/g2").run()?;
    cmd!(sh, "tar -xf sled-common.tar -C {a4x2_dir}/cargo-bay/g3").run()?;
```

This
- creates an intermediate tar of the common dir
- extracts it into each of g0/g1/g2/g3's individual directories

which, since tar does not delete any of the files in `g{0,1,2,3}`, this simply adds them in place in addition to anything else there